### PR TITLE
Better dynamic datetime field sets

### PR DESCRIPTION
**Pull request by sffc**
#1317

### CODE REVIEW

**Comment by sffc** [comment]
So if I change time field sets to be modeled with an option TimePrecision, then I remove a lot of these problems. Instead of H, HM, and HMS, I have just T, and I can make YMD and YMDT first-class types.

```rust
enum TimePrecision { Hour, Minute, Second, SecondF1, SecondF2, ..., SecondF9 }
pub struct T { length, alignment, time_precision }
pub struct ET { length, alignment, time_precision }
pub struct YMDT { length, alignment, year_style, time_precision }
pub struct DateTimeFieldSet { YMDT(fieldset::YMDT), ET(fieldset::ET), ... }
```

And everything cleanly composes.

None of this impacts data loading fortunately, because the only special data is the day period, which is in all of these since all time formats require the hour.

**Comment by sffc** [comment]
Proposal: https://docs.google.com/document/d/1SkxoitlCFiQ_KGW3dmRk7lbumGd_N1rfYJLMlkiXvh4/edit?tab=t.0

**Comment by sffc** [comment]
The last TODO on this is

```rust
impl<DT, Z> GetField<CompositeFieldSet> for Combo<DT, Z> {
    fn get_field(&self) -> CompositeFieldSet {
        todo!()
    }
}
```

Need to think more about how to go about implementing that fn.

**Comment by sffc** [comment]
This is ready for review. The individual commits are not useful and some things were redone in later commits. But there is no data so an all-in-one review should be fine.

I'm working on another commit that regens data and deletes a bunch more code, but I won't push it here since this is a fine standalone PR.

File: components/datetime/tests/simple_test.rs [hunk 1]
```diff
@@ -25,19 +23,19 @@ const EXPECTED_DATETIME: &[&str] = &[
     "शुक्रवार, 22 दिसंबर 2023, 9:22 pm",
     "December 22, 2023, 9:22:53 PM",
     "22 décembre 2023, 21:22:53",
-    "2023年12月22日 21:22:53",
```
Comment (sffc) [inline components/datetime/tests/simple_test.rs:28]: I'm inclined to pull this behavior change out to a separate PR because it isn't really caused by the new dynamic field sets. It is just exposed by them.

File: components/datetime/src/combo.rs [hunk 2]
```diff
@@ -66,232 +69,60 @@ use icu_provider::marker::NeverMarker;
...
+    ///
+    /// ```no_run
+    /// use icu::datetime::fieldset::{self, Combo};
+    /// use icu::datetime::neo_skeleton::NeoSkeletonLength;
+    ///
+    /// let mut field_set: Combo<fieldset::YMD, fieldset::L> = unimplemented!();
+    ///
+    /// field_set.as_mut_dt().length = NeoSkeletonLength::Medium;
+    /// ```
+    pub fn as_mut_dt(&mut self) -> &mut DT {
```
Comment (Manishearth) [inline components/datetime/src/combo.rs:108]: nit (optional): `dt_mut()` or `get_dt_mut()`

thought: also happy with a public field but i get why we want to keep it changeable

File: components/datetime/src/dynamic.rs [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+use crate::neo_serde::*;
+use crate::options::preferences::HourCycle;
+use crate::raw::neo::RawNeoOptions;
+use crate::scaffold::GetField;
+use crate::{fields, fieldset, NeoSkeletonLength};
+use icu_provider::prelude::*;
+
+/// An enumeration over all possible date field sets.
+///
+/// 📏 Caution: This enumeration links more data than the
```
Comment (Manishearth) [inline components/datetime/src/dynamic.rs:15]: nit: "links more data" when? just mentioning the enum clearly doesn't, but it does when used as a fieldset parameter to Formatter. I don't think people will assume that constructing the enum will link in additional data, but they won't necessarily know what the threshold is.

File: components/datetime/src/fieldset.rs [hunk 4]
```diff
@@ -21,26 +24,70 @@ use icu_calendar::{
...
+    ($fallback:expr, $actual:expr) => {
+        $actual
+    };
+    ($fallback:expr,) => {
+        $fallback
+    };
+}
+
+macro_rules! ternary {
+    ($present:expr, $missing:expr, yes) => {
```
Comment (Manishearth) [inline components/datetime/src/fieldset.rs:56]: nit: should we match `no`?

File: components/datetime/src/fieldset.rs [hunk 5]
```diff
@@ -153,6 +228,76 @@ macro_rules! impl_marker_with_options {
...
+    (
+        $type:ident,
+        $composite:ident,
+        $enum:ident,
+        $wrap:ident
+    ) => {
+        impl $type {
+            /// Associates this field set with a specific non-location format time zone, as in
+            /// “Pacific Daylight Time”.
+            pub fn z(self) -> Combo<Self, Zs> {
```
Comment (Manishearth) [inline components/datetime/src/fieldset.rs:275]: praise: I like this idea

**Review by Manishearth — COMMENTED** [review]
I'll have a second look tomorrow but I like the overall direction here!

File: components/datetime/src/fieldset.rs (hunk 4 above)
Comment (sffc) [reply components/datetime/src/fieldset.rs:56]: We match either "present" or "missing", i.e. "yes" and ""

File: components/datetime/tests/simple_test.rs (hunk 1 above)
Comment (sffc) [reply components/datetime/tests/simple_test.rs:28]: I tried splitting it, but it was more work than I thought was needed, because I would need to re-write this test with the old neo skeleton API only to immediately replace it with the new dynamic field set API.

File: components/datetime/src/combo.rs [hunk 6]
```diff
@@ -66,232 +72,67 @@ use icu_provider::marker::NeverMarker;
...
+/// ```
+/// use icu::datetime::fieldset::{YMD, dynamic::DateFieldSet};
+/// use icu::datetime::DateTimeFormatter;
+/// use icu::locale::locale;
+/// use icu::timezone::IxdtfParser;
+/// use writeable::assert_try_writeable_eq;
+///
+/// let formatter = DateTimeFormatter::try_new(
+///     &locale!("en-US").into(),
+///     DateFieldSet::YMD(YMD::long()).v(),
```
Comment (Manishearth) [inline components/datetime/src/combo.rs:88]: nit: comment saying that this is Combo<YMD, V>

**Review by Manishearth — APPROVED** [review]
Feel free to fix the nit in one of your stacks

File: components/datetime/src/combo.rs (hunk 6 above)
Comment (sffc) [reply components/datetime/src/combo.rs:88]: Will address this in #5807
